lib/std/os/uefi/bits.zig

```diff
@@ -0,0 +1,471 @@
+const std = @import("../../std.zig");
+
+const time = std.time;
+
+/// The calling convention used for all external functions part of the UEFI API.
+pub const cc = switch (@import("builtin").target.cpu.arch) {
+    .x86_64 => .Win64,
+    else => .C,
+};
+
+/// A pointer in physical address space.
+pub const PhysicalAddress = u64;
+
+/// A pointer in virtual address space.
+pub const VirtualAddress = u64;
+
+/// A logical block address
+pub const LogicalBlockAddress = u64;
+
+/// An EFI Handle represents a collection of related interfaces.
+pub const Handle = *const opaque {};
+
+/// A handle to an event structure.
+pub const Event = *const opaque {};
+
+/// File Handle as specified in the EFI Shell Spec
+pub const FileHandle = *const opaque {};
+
+pub const Crc32 = std.hash.crc.Crc32IsoHdlc;
+
+pub const MacAddress = extern struct {
+    address: [32]u8,
+};
+
+pub const Ipv4Address = extern struct {
+    address: [4]u8,
+};
+
+pub const Ipv6Address = extern struct {
+    address: [16]u8,
+};
+
+pub const IpAddress = extern union {
+    v4: Ipv4Address,
+    v6: Ipv6Address,
+};
+
+/// This structure represents time information.
+pub const Time = extern struct {
+    /// 1900 - 9999
+    year: u16,
+
+    /// 1 - 12
+    month: u8,
+
+    /// 1 - 31
+    day: u8,
+
+    /// 0 - 23
+    hour: u8,
+
+    /// 0 - 59
+    minute: u8,
+
+    /// 0 - 59
+    second: u8,
+
+    /// 0 - 999999999
+    nanosecond: u32,
+
+    /// The time's offset in minutes from UTC.
+    /// Allowed values are -1440 to 1440 or unspecified_timezone
+    timezone: i16,
+    daylight: packed struct {
+        _pad1: u6,
+
+        /// If true, the time has been adjusted for daylight savings time.
+        in_daylight: bool,
+
+        /// If true, the time is affected by daylight savings time.
+        adjust_daylight: bool,
+    },
+
+    /// Time is to be interpreted as local time
+    pub const unspecified_timezone: i16 = 0x7ff;
+
+    /// Returns the time in seconds since 1900-01-01 00:00:00.
+    pub fn toEpochSeconds(self: Time) u64 {
+        var year: u16 = 1900;
+        var days: u32 = 0;
+
+        while (year < self.year) : (year += 1) {
+            days += time.epoch.getDaysInYear(year);
+        }
+
+        var month: u8 = 1;
+        while (month < self.month) : (month += 1) {
+            const leap_kind: time.epoch.YearLeapKind = if (time.epoch.isLeapYear(self.year)) .leap else .not_leap;
+
+            days += time.epoch.getDaysInMonth(leap_kind, @enumFromInt(month));
+        }
+
+        days += self.day - 1;
+
+        return days * time.s_per_day +
+            @as(u32, self.hour) * time.s_per_hour +
+            @as(u16, self.minute) * time.s_per_min +
+            self.second;
+    }
+
+    /// Returns the time in nanoseconds since 1900-01-01 00:00:00.
+    pub fn toEpochNanoseconds(self: Time) u128 {
```

Comment:
This method is causing an integer overflow panic.
```zig
const std = @import("std");

pub fn main() !void {
    const stat = try std.posix.fstat(std.fs.cwd().fd);

    std.debug.print("{} ", .{stat.atim});
    std.debug.print("{}\n", .{stat.atime()});

    std.debug.print("{} ", .{stat.mtim});
    std.debug.print("{}\n", .{stat.mtime()});

    std.debug.print("{} ", .{stat.ctim});
    std.debug.print("{}\n", .{stat.ctime()});
}
```
```
os.uefi.bits.Time{ .year = 2024, .month = 3, .day = 31, .hour = 0, .minute = 0, .second = 0, .nanosecond = 0, .timezone = 2047, .daylight = os.uefi.bits.Time.Time__struct_1282{ ._pad1 = 0, .in_daylight = false, .adjust_daylight = false } } 3920832000
os.uefi.bits.Time{ .year = 2024, .month = 3, .day = 31, .hour = 22, .minute = 28, .second = 8, .nanosecond = 0, .timezone = 2047, .daylight = os.uefi.bits.Time.Time__struct_1282{ ._pad1 = 0, .in_daylight = false, .adjust_daylight = false } } 3920912888
os.uefi.bits.Time{ .year = 1980, .month = 0, .day = 0, .hour = 0, .minute = 0, .second = 0, .nanosecond = 0, .timezone = 2047, .daylight = os.uefi.bits.Time.Time__struct_1282{ ._pad1 = 0, .in_daylight = false, .adjust_daylight = false } } panic: integer overflow 3186494919
```

Reply:
Correction: this is with `toEpochSeconds` and not `toEpochNanoseconds`

Reply:
My suspicion would be `days += self.day - 1;`. Based on the doc comment for `day`, it should be between 1 and 31, but you're passing a struct with `.day = 0`, so something looks to be out of sync.

Reply:
Yeah, I am not sure why it is like that. My implementation didn't do this (the one currently in upstream). This comes from @truemedian's branch which I incorporated.

Reply:
Going further with this test program, I get this output:
```
os.uefi.bits.Time{ .year = 2024, .month = 3, .day = 31, .hour = 0, .minute = 0, .second = 0, .nanosecond = 0, .timezone = 2047, .daylight = os.uefi.bits.Time.Time__struct_1283{ ._pad1 = 0, .in_daylight = false, .adjust_daylight = false } } 3920832000
os.uefi.bits.Time{ .year = 2024, .month = 3, .day = 31, .hour = 22, .minute = 51, .second = 46, .nanosecond = 0, .timezone = 2047, .daylight = os.uefi.bits.Time.Time__struct_1283{ ._pad1 = 0, .in_daylight = false, .adjust_daylight = false } } 3920914306
os.uefi.bits.Time{ .year = 1980, .month = 0, .day = 0, .hour = 0, .minute = 0, .second = 0, .nanosecond = 0, .timezone = 2047, .daylight = os.uefi.bits.Time.Time__struct_1283{ ._pad1 = 0, .in_daylight = false, .adjust_daylight = false } }
Days in years: 29219

Days in months: 29219
panic: integer overflow 3186465173
Panicked during a panic. Aborting
```

```zig
const std = @import("std");

fn toEpochSeconds(self: std.os.uefi.bits.Time) u64 {
    var year: u16 = 1900;
    var days: u32 = 0;

    while (year < self.year) : (year += 1) {
        days += std.time.epoch.getDaysInYear(year);
    }
    std.debug.print("\nDays in years: {}\n", .{days});

    var month: u8 = 1;
    while (month < self.month) : (month += 1) {
        const leap_kind: std.time.epoch.YearLeapKind = if (std.time.epoch.isLeapYear(self.year)) .leap else .not_leap;

        days += std.time.epoch.getDaysInMonth(leap_kind, @enumFromInt(month));
    }
    std.debug.print("\nDays in months: {}\n", .{days});

    days += self.day - 1;
    std.debug.print("Days: {}\n", .{days});

    const hours = @as(u32, self.hour) * std.time.s_per_hour;
    std.debug.print("Hours: {}\n", .{hours});

    const minutes = @as(u16, self.minute) * std.time.s_per_min;
    std.debug.print("Minutes: {}\n", .{minutes});

    const days_as_sec = days * std.time.s_per_day;
    std.debug.print("Days as seconds: {}\n", .{days_as_sec});

    return days_as_sec +
        hours +
        minutes +
        self.second;
}

pub fn main() !void {
    const stat = try std.posix.fstat(std.fs.cwd().fd);

    std.debug.print("{} ", .{stat.atim});
    std.debug.print("{}\n", .{stat.atime()});

    std.debug.print("{} ", .{stat.mtim});
    std.debug.print("{}\n", .{stat.mtime()});

    std.debug.print("{} ", .{stat.ctim});
    std.debug.print("{}\n", .{toEpochSeconds(stat.ctim)});
}
```
So I think @ianprime0509 is right on this one.

Reply:
Seems like my original implementation that's in master breaks with the atime.
```
os.uefi.bits.Time{ .year = 2024, .month = 3, .day = 31, .hour = 0, .minute = 0, .second = 0, .nanosecond = 0, .timezone = 2047, .daylight = os.uefi.bits.Time.Time__struct_1283{ ._pad1 = 0, .in_daylight = false, .adjust_daylight = false } } panic: integer overflow 3186491235
Panicked during a panic. Aborting.
```

Reply:
Ok, I got further to the second to ns conversion but now casting from u64 to u128 is yielding a `panic: integer cast truncated bits`.

Reply:
Based on https://uefi.org/specs/UEFI/2.10/Apx_A_GUID_and_Time_Formats.html the UEFI epoch is 1582-10-15. I've fixed this in #19549.

Reply:
This is the wrong value. The UEFI GUID Timestamp uses the 1582 epoch, but GUID values are hardcoded and this really doesn't affect UEFI code. The EFI_TIME epoch is Jan 1 1900.

Reply:
Yeah, though when we convert to UNIX time, it actually doesn't work because the epoch value in std is 0 so we are doing `epoch - 0`. But even without that subtraction, converting the u64 to a u128 is causing an int truncated panic.

Reply:
I would also argue that std.os.uefi definitely does not necessitate the introduction of an entire DateTime handling system into std. UEFI's `Time` needs only 2 (or 3) functions:
- convert the Time into seconds since the UEFI epoch, Jan 1 1900, which is just a few loops.
- convert the Time into nanoseconds since the UEFI epoch, uses the above plus a bit of math
- convert the Time into nanoseconds since the UNIX epoch, which is the epoch the rest of zig uses.

The first two could easily be inlined into the last, but skipping the final offset is less math, and therefore just a bit faster for use cases that don't require a specific epoch (like Instant).

Calculating the seconds from the UNIX epoch to the UEFI epoch is as simple as creating a UEFI Time object of the UNIX epoch and asking how many seconds it was since the UNIX epoch.

There is a fully functional implementation of this, with no overflows, on [my branch](https://github.com/truemedian/zig/blob/uefi-rework/lib/std/os/uefi/bits.zig#L88-L133).

Reply:
How do you get no overflows? This PR is based on your branch and I only had to change 1 line because of a const that doesn't exist.

Reply:
> The EFI_TIME epoch is Jan 1 1900.

I've fixed it and added a comment on my branch. That's confusing!

> I would also argue that std.os.uefi definitely does not necessitate the introduction of an entire DateTime handling system into std.

No, but I think it useful for your `DateTime` struct. It also gives you EpochSeconds -> DateTime. Since the struct layout is equivalent (except for your flags at the end), you could even use the `DateTime` struct inside yours.

> - convert the Time into seconds since the UEFI epoch, Jan 1 1900, which is just a few loops.

While it may not matter for your needs, branching on leap year and month number in your loop is is slow. You can get nearly a 2-3x improvement with the [Cassio and Neri implementation](https://onlinelibrary.wiley.com/doi/epdf/10.1002/spe.3172).

Reply:
> How do you get no overflows?

@truemedian chose a `u32` to hold the seconds. Since the years is a `u16`, `366 * std.math.maxInt(u16) < std.math.maxInt(u32)` and the years can't overflow.

Reply:
@clickingbuttons I do as well, however @truemedian's implementation is kinda based on mine which exists in std. The problem I'm getting with their implementation is converting the u64 time to the u128 time for UNIX. This triggers a `panic: integer cast truncated bits 3186670993`. This doesn't make sense as u128 is bigger than u64. What this panic is basically saying is that it is truncating a u64 to fit in a u128.

Reply:
> If you give a full stack trace I can help debug.

Word of warning: we still cannot produce stack traces on UEFI, we're missing the information required to walk the stack backwards.

Reply:
I'm not even int casting. @nektro has tried helping me with this problem as well on Discord. I traced the issue to the `@as(u128, v)` call with print calls.

Reply:
Hm, that's very strange. Is there a way to run behavior and stdlib tests on UEFI?

Reply:
Kinda, however there are more important issues we need to get through first. Like zig actually being able to find itself in UEFI.